File: app/src/main/java/kr/zagros/shwan/moviemvvm/DataSource/PagedSources/MoviesDataSourceFactory.java
package kr.zagros.shwan.moviemvvm.DataSource.PagedSources;

import androidx.lifecycle.MutableLiveData;
import androidx.paging.DataSource;
import android.util.Log;

import java.util.concurrent.Executor;

import kr.zagros.shwan.moviemvvm.DataSource.Client.ApiService;

public class MoviesDataSourceFactory extends DataSource.Factory {
    private static final String TAG = "MoviesDataSourceFactory";

    private MoviesDataSource moviesDataSource;
    private MutableLiveData<MoviesDataSource> mutableLiveData;

    private ApiService apiService;
    private Executor executor;

    public MoviesDataSourceFactory(ApiService apiService, Executor executor) {
        this.apiService = apiService;
        this.executor = executor;
        this.mutableLiveData = new MutableLiveData<>();
    }

    @Override
    public DataSource create() {
        Log.e(TAG, "create: " );
        moviesDataSource=new MoviesDataSource(executor,apiService);
        mutableLiveData.postValue(moviesDataSource);
        return moviesDataSource;
    }

    public MutableLiveData<MoviesDataSource> getMutableLiveData() {
        return mutableLiveData;
    }
}

File: app/src/main/java/kr/zagros/shwan/moviemvvm/DataSource/PagedSources/MoviesDataSourceFactory.kt
package kr.zagros.shwan.moviemvvm.DataSource.PagedSources

import androidx.lifecycle.MutableLiveData
import androidx.paging.DataSource
import android.util.Log

import java.util.concurrent.Executor

import kr.zagros.shwan.moviemvvm.DataSource.Client.ApiService
import kr.zagros.shwan.moviemvvm.Entities.Movie

class MoviesDataSourceFactory(private val apiService: ApiService, private val executor: Executor) : DataSource.Factory<Long, Movie>() {

    private var moviesDataSource: MoviesDataSource? = null
    val mutableLiveData: MutableLiveData<MoviesDataSource> = MutableLiveData()

    override fun create(): DataSource<Long, Movie> {
        Log.e(TAG, "create: ")
        moviesDataSource = MoviesDataSource(executor, apiService)
        mutableLiveData.postValue(moviesDataSource)
        return moviesDataSource as MoviesDataSource
    }

    companion object {
        private const val TAG = "MoviesDataSourceFactory"
    }
}
